### PR TITLE
feat: add function to get cluster migration info

### PR DESCRIPTION
**Pull request by levy5307**
add a function to get info used for cluster load balance

### CODE REVIEW

**Review by hycdong — APPROVED** [review]

File: src/meta/greedy_load_balancer.cpp [hunk 1]
```diff
@@ -39,6 +39,44 @@ namespace replication {
...
+            count = ns.primary_count();
+        }
+        break;
+    default:
+        break;
+    }
+    return (uint32_t)count;
+}
+
+uint32_t get_skew(const std::map<rpc_address, uint32_t> &count_map)
```
Comment (Shuo-Jia) [inline src/meta/greedy_load_balancer.cpp:66]: `get_diff` or `get_range`?

File: src/meta/greedy_load_balancer.h [hunk 2]
```diff
@@ -150,9 +153,75 @@ class greedy_load_balancer : public simple_load_balancer
...
-    bool cluster_replica_balance(const cluster_balance_type type);
+    bool cluster_replica_balance(const meta_view *global_view, const cluster_balance_type type);
+
+    bool do_cluster_replica_balance(const meta_view *global_view, const cluster_balance_type type);
+
+    struct app_migration_info
+    {
+        int32_t app_id;
+        std::string app_name;
+        std::vector<std::map<rpc_address, partition_status::type>> partitions;
```
Comment (Shuo-Jia) [inline src/meta/greedy_load_balancer.h:164]: base your code, vector `index` means `pid`?

https://github.com/XiaoMi/rdsn/pull/866/files#diff-c013463d4991b89fb5d5328eb8894552e0c655d1973fa800a0bb2a5e9d347c0aR208

File: src/meta/greedy_load_balancer.h [hunk 3]
```diff
@@ -150,9 +153,75 @@ class greedy_load_balancer : public simple_load_balancer
...
+                }
+            }
+            return partition_status::PS_INACTIVE;
+        }
+    };
+
+    struct node_migration_info
+    {
+        rpc_address address;
+        std::map<std::string, partition_set> partitions;
```
Comment (Shuo-Jia) [inline src/meta/greedy_load_balancer.h:190]: suggest all `map` add comment to show the  k-v means

File: src/meta/greedy_load_balancer.h [hunk 4]
```diff
@@ -150,9 +153,75 @@ class greedy_load_balancer : public simple_load_balancer
...
+        partition_set future_partitions;
+        bool operator<(const node_migration_info &another) const
+        {
+            if (address < another.address)
```
Comment (Shuo-Jia) [inline src/meta/greedy_load_balancer.h:194]:
```suggestion
            return address < another.address;
```

other same

**Review by Smityz — APPROVED** [review]

File: src/meta/greedy_load_balancer.cpp [hunk 5]
```diff
@@ -963,10 +1000,120 @@ bool greedy_load_balancer::cluster_replica_balance(const cluster_balance_type ty
...
+            return false;
+        }
+    }
+
+    const app_mapper &all_apps = *global_view->apps;
+    app_mapper apps;
+    for (const auto &kv : all_apps) {
+        const std::shared_ptr<app_state> &app = kv.second;
+        if (is_ignored_app(app->app_id) || app->is_bulk_loading || app->splitting()) {
+            return false;
```
Comment (Shuo-Jia) [inline src/meta/greedy_load_balancer.cpp:1034]: why meet `ignored_app` is `return`? actually should `continue`?

File: src/meta/greedy_load_balancer.cpp (hunk 1 above)
Comment (levy5307) [reply src/meta/greedy_load_balancer.cpp:66]: skew is more accurate

File: src/meta/greedy_load_balancer.cpp [hunk 6]
```diff
@@ -39,6 +39,44 @@ namespace replication {
...
 
+uint32_t get_count(const node_state &ns, cluster_balance_type type, int32_t app_id)
+{
+    unsigned count = 0;
+    switch (type) {
+    case cluster_balance_type::Secondary:
+        if (app_id > 0) {
+            count = ns.partition_count(app_id) - ns.primary_count(app_id);
+        } else {
+            count = ns.partition_count() - ns.primary_count();
```
[the remark this answers is not in the document]
Comment (levy5307) [reply src/meta/greedy_load_balancer.cpp:50]: It's too complex to use  ternary expression. I don't think it's a good idea

File: src/meta/greedy_load_balancer.cpp [hunk 7]
```diff
@@ -39,6 +39,44 @@ namespace replication {
 DSN_DEFINE_bool("meta_server", balance_cluster, false, "whether to enable cluster balancer");
 DSN_TAG_VARIABLE(balance_cluster, FT_MUTABLE);
 
+uint32_t get_count(const node_state &ns, cluster_balance_type type, int32_t app_id)
```
[the remark this answers is not in the document]
Comment (levy5307) [reply src/meta/greedy_load_balancer.cpp:42]: OK I will change it in next pull request.

File: src/meta/greedy_load_balancer.cpp (hunk 5 above)
Comment (levy5307) [reply src/meta/greedy_load_balancer.cpp:1034]: yes, you are right

File: src/meta/greedy_load_balancer.h (hunk 2 above)
Comment (levy5307) [reply src/meta/greedy_load_balancer.h:164]: yes

**Pull request merged**